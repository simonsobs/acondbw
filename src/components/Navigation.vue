<template>
  <v-list shaped>
    <v-list-item
      link
      router
      v-for="edge in edges"
      :key="edge.node.typeId"
      :to="'/' + edge.node.name"
    >
      <v-list-item-action>
        <v-icon v-text="edge.node.icon"></v-icon>
      </v-list-item-action>
      <v-list-item-content>
        <v-list-item-title v-text="edge.node.plural" class="capitalize"></v-list-item-title>
      </v-list-item-content>
    </v-list-item>
  </v-list>
</template>

<script>
import ALL_PRODUCTS_TYPES from "@/graphql/AllProductTypes.gql";

export default {
  name: "Navigation",
  data: () => ({
    edges: null,
    error: null
  }),
  apollo: {
    edges: {
      query: ALL_PRODUCTS_TYPES,
      update: data =>
        data.allProductTypes ? data.allProductTypes.edges : null,
      result(result) {
        this.error = result.error ? result.error : null;
      }
    }
  }
};
</script>
