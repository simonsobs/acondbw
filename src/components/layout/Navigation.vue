<template>
  <div class="navigation">
    <v-list v-if="loaded" shaped>
      <v-list-item
        link
        router
        v-for="edge in edges"
        :key="edge.node.typeId"
        :to="{
          name: 'ProductList',
          params: { productTypeName: edge.node.name },
        }"
      >
        <v-list-item-action class="mr-5">
          <v-icon v-text="edge.node.icon"></v-icon>
        </v-list-item-action>
        <v-list-item-content>
          <v-list-item-title
            v-text="edge.node.plural"
            class="capitalize condensed-font font-weight-medium"
          ></v-list-item-title>
        </v-list-item-content>
        <v-list-item-icon class="ml-2">
          <v-chip
            small
            v-if="!!edge.node.products.totalCount"
            v-text="edge.node.products.totalCount"
          ></v-chip>
        </v-list-item-icon>
      </v-list-item>
    </v-list>
    <v-alert v-else-if="empty" outlined dense type="info" class="ma-2">
      No product types are defined.
    </v-alert>
    <v-progress-circular
      v-else-if="loading"
      indeterminate
      :size="18"
      :width="3"
      color="secondary"
      class="mx-5 mt-5"
    ></v-progress-circular>
    <v-alert v-else-if="error" outlined dense type="error" class="ma-2">
      {{ error }}
    </v-alert>
    <v-bottom-navigation absolute class="px-3 justify-start align-center">
      <span class="grey--text text-caption">
        v{{ $store.getters.appVersion }}
      </span>
      <!-- <v-spacer></v-spacer>
      <v-icon>mdi-plus-thick</v-icon> -->
    </v-bottom-navigation>
    <dev-tool-loading-state-overriding-menu
      v-model="devtoolState"
    ></dev-tool-loading-state-overriding-menu>
  </div>
</template>

<script>
import ALL_PRODUCT_TYPES from "@/graphql/queries/AllProductTypes.gql";

import State from "@/utils/LoadingState.js";
import DevToolLoadingStateOverridingMenu from "@/components/utils/DevToolLoadingStateOverridingMenu.vue";

export default {
  name: "Navigation",
  components: {
    DevToolLoadingStateOverridingMenu,
  },
  data: () => ({
    init: true,
    edges: null,
    error: null,
    devtoolState: null,
    State: State,
  }),
  computed: {
    state() {
      if (this.devtoolState) return this.devtoolState;
      if (this.$apollo.loading) return State.LOADING;
      if (this.error) return State.ERROR;
      if (this.edges) {
        if (this.edges.length) return State.LOADED;
        return State.EMPTY;
      }
      if (this.init) return State.INIT;
      return State.NONE;
    },
    loading() {
      return this.state == State.LOADING;
    },
    loaded() {
      return this.state == State.LOADED;
    },
    empty() {
      return this.state == State.EMPTY;
    },
    notFound() {
      return this.state == State.NONE;
    },
  },
  watch: {
    devtoolState: function () {
      if (this.devtoolState) {
        this.init = this.devtoolState == State.INIT;
      }
      this.error =
        this.devtoolState == State.ERROR ? "Error from Dev Tools" : null;
    },
    "$store.state.nApolloMutations": function () {
      this.$apollo.queries.edges.refetch();
    },
  },
  apollo: {
    edges: {
      query: ALL_PRODUCT_TYPES,
      update: (data) =>
        data.allProductTypes ? data.allProductTypes.edges : null,
      result(result) {
        this.init = false;
        this.error = result.error || null;
      },
    },
  },
};
</script>
