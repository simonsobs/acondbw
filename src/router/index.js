import Vue from "vue";
import VueRouter from "vue-router";
import VueMeta from "vue-meta";

import Frame from "@/components/layout/Frame";
import NullFrame from "@/components/layout/NullFrame";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue"

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import store from "@/store";

Vue.use(VueRouter);
Vue.use(VueMeta);

const routes = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const signedIn = !!store.state.auth.token;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: {
      default: OAuthRedirect,
      frame: NullFrame,
    },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: {
      default: SignIn,
      frame: NullFrame,
    },
  },
  {
    path: "/auth",
    name: "Auth",
    components: {
      default: Auth,
      frame: NullFrame,
    },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: {
      default: SignInError,
      frame: NullFrame,
    },
  },
  {
    path: "/admin/scratch",
    name: "AdminScratch",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/Scratch"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken"
        ),
      frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: {
      default: () =>
        import(
          /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError"
        ),
        frame: () =>
        import(
          /* webpackChunkName: "admin" */ "@/components/layout/FrameAdmin"
        ),
    },
  },
  {
    path: "/about",
    name: "About",
    component: () =>
      import(/* webpackChunkName: "about" */ "@/views/framework/About.vue"),
  },
  {
    path: "/add-product/:productTypeName",
    components: {
      default: ProductAdd,
      frame: NullFrame,
    },
    meta: { requiresAuth: true }
  },
  {
    path: "/product/:productTypeName",
    components: {
      default: ProductTop,
      frame: Frame,
    },
    meta: { requiresAuth: true },
    children: [
      {
        path: "",
        name: "ProductList",
        component: ProductList,
      },
      {
        path: ":name",
        name: "ProductItem",
        component: ProductItem,
      },
    ],
  },
  {
    path: "*",
    name: "NotFound",
    components: {
      default: () =>
        import(/* webpackChunkName: "error" */ "@/views/framework/NotFound.vue"),
        frame: NullFrame,
    },
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes,
});

router.beforeEach((to, from, next) => {
  const authRequired = to.matched.some((record) => record.meta.requiresAuth);
  const signedIn = !!store.state.auth.token;
  if (authRequired) {
    if (signedIn) {
      next();
    } else {
      next("/");
    }
  } else {
    next();
  }
});

export default router;
