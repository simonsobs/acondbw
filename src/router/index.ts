import Vue from "vue";
import VueRouter, { RouteConfig } from "vue-router";
import VueMeta from "vue-meta";

import { default as defaultPinia } from "@/stores";
import { useAuthStore } from "@/stores/auth";

import Frame from "@/components/layout/Frame.vue";
import NullFrame from "@/components/layout/NullFrame.vue";

import Home from "@/views/framework/Home.vue";
import Entry from "@/views/framework/Entry.vue";
import SearchTest from "@/views/framework/SearchTest.vue";

import OAuthRedirect from "@/views/auth/OAuthRedirect.vue";

import SignIn from "@/views/auth/SignIn.vue";
import Auth from "@/views/auth/Auth.vue";
import SignInError from "@/views/auth/SignInError.vue";

import ProductTop from "@/views/product/ProductTop.vue";
import ProductList from "@/views/product/ProductList.vue";
import ProductItem from "@/views/product/ProductItem.vue";
import ProductAdd from "@/views/product/ProductAdd.vue";

import ProductType from "@/views/admin/ProductType.vue";

const About = () => import("@/views/framework/About.vue");
const NotFound = () => import("@/views/framework/NotFound.vue");
const AccessDenied = () => import("@/views/framework/AccessDenied.vue");

const FrameAdmin = () =>
  import(/* webpackChunkName: "admin" */ "@/components/admin/FrameAdmin.vue");
const Version = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Version.vue");
const Log = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Log.vue");
const Config = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Config.vue");
const Theme = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/Theme.vue");
const User = () =>
  import(/* webpackChunkName: "admin" */ "@/views/admin/User.vue");
const AdminAppToken = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppToken.vue"
  );
const AdminAppAuth = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppAuth.vue"
  );
const AdminAppTokenError = () =>
  import(
    /* webpackChunkName: "admin" */ "@/views/admin-token/AdminAppTokenError.vue"
  );

Vue.use(VueRouter);
Vue.use(VueMeta);

let pinia = defaultPinia;

function setPinia(val) {
  // because not sure how to monkey patch pinia for tests
  pinia = val;
}

function setDefaultPinia() {
  setPinia(defaultPinia);
}

const routes: Array<RouteConfig> = [
  {
    path: "/",
    name: "Entry",
    components: {
      default: Entry,
      frame: NullFrame,
    },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next({ name: "Dashboard" });
      } else {
        next();
      }
    },
  },
  {
    path: "/dashboard",
    name: "Dashboard",
    components: {
      default: Home,
      frame: Frame,
    },
    // meta: { requiresAuth: true },
    beforeEnter: (to, from, next) => {
      const auth = useAuthStore(pinia);
      const signedIn = auth.isSignedIn;
      if (signedIn) {
        next();
      } else {
        next({ name: "Entry" });
      }
    },
  },
  {
    path: "/search-test",
    name: "SearchTest",
    components: { default: SearchTest, frame: Frame },
    meta: { requiresAuth: true },
  },
  {
    path: "/oauth-redirect",
    name: "OAuthRedirect",
    components: { default: OAuthRedirect, frame: NullFrame },
  },
  {
    path: "/signin",
    name: "SignIn",
    components: { default: SignIn, frame: NullFrame },
  },
  {
    path: "/auth",
    name: "Auth",
    components: { default: Auth, frame: NullFrame },
  },
  {
    path: "/signin-error",
    name: "SignInError",
    components: { default: SignInError, frame: NullFrame },
  },
  {
    path: "/admin",
    name: "Admin",
    redirect: { name: "AdminVersion" },
  },
  {
    path: "/admin/product-type",
    name: "ProductType",
    components: { default: ProductType, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/versions",
    name: "AdminVersion",
    components: { default: Version, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/log",
    name: "AdminLog",
    components: { default: Log, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/config",
    name: "AdminConfig",
    components: { default: Config, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/theme",
    name: "AdminTheme",
    components: { default: Theme, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/users",
    name: "AdminUser",
    components: { default: User, frame: FrameAdmin },
    meta: { requiresAdmin: true },
  },
  {
    path: "/admin/token",
    name: "AdminAppToken",
    components: { default: AdminAppToken, frame: FrameAdmin },
  },
  {
    path: "/admin/auth",
    name: "AdminAppAuth",
    components: { default: AdminAppAuth, frame: FrameAdmin },
  },
  {
    path: "/admin/token-error",
    name: "AdminAppTokenError",
    components: { default: AdminAppTokenError, frame: FrameAdmin },
  },
  {
    path: "/about",
    name: "About",
    component: About,
  },
  {
    path: "/add-product/:productTypeName",
    name: "ProductAdd",
    components: { default: ProductAdd, frame: NullFrame },
    meta: { requiresAuth: true },
  },
  {
    path: "/product/:productTypeName",
    components: { default: ProductTop, frame: Frame },
    meta: { requiresAuth: true },
    children: [
      { path: "", name: "ProductList", component: ProductList },
      { path: ":name", name: "ProductItem", component: ProductItem },
    ],
  },
  {
    path: "/access-denied",
    name: "AccessDenied",
    components: { default: AccessDenied, frame: NullFrame },
  },
  {
    path: "*",
    name: "NotFound",
    components: { default: NotFound, frame: NullFrame },
  },
];

function createRouter() {
  const router = new VueRouter({
    mode: "history",
    base: process.env.BASE_URL,
    routes,
  });

  router.beforeEach((to, from, next) => {
    const auth = useAuthStore(pinia);

    const adminRequired = to.matched.some(
      (record) => record.meta.requiresAdmin
    );
    const isAdmin = auth.isAdmin;
    if (adminRequired && !isAdmin) {
      next({ name: "AccessDenied" });
      return;
    }

    const authRequired = to.matched.some((record) => record.meta.requiresAuth);
    const isSignedIn = auth.isSignedIn as boolean;
    if (authRequired && !isSignedIn) {
      next({ name: "AccessDenied" });
      return;
    }
    next();
  });

  return router;
}

const router = createRouter();

function checkAuthForCurrentRoute() {
  const authRequired = router.currentRoute.matched.some(
    (record) => record.meta.requiresAuth
  );
  const auth = useAuthStore(pinia);
  const signedIn = auth.isSignedIn as boolean;
  if (authRequired && !signedIn) {
    router.push("/");
  }
}

export { createRouter, checkAuthForCurrentRoute, setPinia, setDefaultPinia };
