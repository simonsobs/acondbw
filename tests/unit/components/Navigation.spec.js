import Vue from "vue";
import VueRouter from "vue-router";
import Vuetify from "vuetify";
import { mount, shallowMount, createLocalVue } from "@vue/test-utils";

import Navigation from "@/components/Navigation";
import router from "@/router";

Vue.use(Vuetify);
Vue.use(VueRouter);

describe("App.vue", () => {
  let localVue;
  let vuetify;
  let wrapper;

  const edges = [
    {
      node: {
        id: "UHJvZHVjdFR5cGU6Mw==",
        typeId: "3",
        name: "simulation",
        order: 1,
        indefArticle: "a",
        singular: "simulation",
        plural: "simulations",
        icon: "mdi-creation",
      },
    },
    {
      node: {
        id: "UHJvZHVjdFR5cGU6MQ==",
        typeId: "1",
        name: "map",
        order: 2,
        indefArticle: "a",
        singular: "map",
        plural: "maps",
        icon: "mdi-map",
      },
    },
    {
      node: {
        id: "UHJvZHVjdFR5cGU6Mg==",
        typeId: "2",
        name: "beam",
        order: 3,
        indefArticle: "a",
        singular: "beam",
        plural: "beams",
        icon: "mdi-spotlight-beam",
      },
    },
  ];

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    wrapper = shallowMount(Navigation, {
      localVue,
      vuetify,
      router,
      mocks: {
        $apollo: {
          queries: {
            edges: {
              loading: false,
            },
          },
        },
      },
    });
    wrapper.setData({
      edges: edges,
    });
  });

  it("match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
