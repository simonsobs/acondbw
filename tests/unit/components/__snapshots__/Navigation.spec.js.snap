// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue match snapshot 1`] = `
<v-list-stub shaped="true" tag="div">
  <v-list-item-stub activeclass="" link="true" to="/simulation" tag="div" router="">
    <v-list-item-action-stub>
      <v-icon-stub></v-icon-stub>
    </v-list-item-action-stub>
    <v-list-item-content-stub>
      <v-list-item-title-stub class="capitalize"></v-list-item-title-stub>
    </v-list-item-content-stub>
  </v-list-item-stub>
  <v-list-item-stub activeclass="" link="true" to="/map" tag="div" router="">
    <v-list-item-action-stub>
      <v-icon-stub></v-icon-stub>
    </v-list-item-action-stub>
    <v-list-item-content-stub>
      <v-list-item-title-stub class="capitalize"></v-list-item-title-stub>
    </v-list-item-content-stub>
  </v-list-item-stub>
  <v-list-item-stub activeclass="" link="true" to="/beam" tag="div" router="">
    <v-list-item-action-stub>
      <v-icon-stub></v-icon-stub>
    </v-list-item-action-stub>
    <v-list-item-content-stub>
      <v-list-item-title-stub class="capitalize"></v-list-item-title-stub>
    </v-list-item-content-stub>
  </v-list-item-stub>
</v-list-stub>
`;
